src/blocks/battery.rs

```diff
@@ -70,274 +75,335 @@ impl PowerSupplyDevice {
+                    if !allow_missing {
+                        return Err(e);
+                    }
+                }
 
-        Ok(device)
+                Ok(device)
+            }
+            Err(e) => Err(BlockError(
+                "battery".into(),
+                format!("Failed to parse Regex {}: {}", device_regex, e),
+            )),
+        }
     }
 }
 
 impl BatteryDevice for PowerSupplyDevice {
     fn is_available(&self) -> bool {
-        // in case of human interface devices (scope=="Device")
-        // we don't check for present==1 because the device subdirectory
-        // being present already implies that the device is available
-        let path = self.device_path.join("scope");
-        if path.exists() && read_file("battery", path).map_or(false, |x| x == "Device") {
-            // device is a human interface device
-            return true;
-        }
-        // normal battery device -> check for present==1
-        let path = self.device_path.join("present");
-        if path.exists() {
-            return read_file("battery", path).map_or(false, |x| x == "1");
+        if let Some(device_path) = &self.device_path {
+            // in case of human interface devices (scope=="Device")
+            // we don't check for present==1 because the device subdirectory
+            // being present already implies that the device is available
+            let path = device_path.join("scope");
+            if path.exists() && read_file("battery", path).map_or(false, |x| x == "Device") {
+                // device is a human interface device
+                return true;
+            }
+            // normal battery device -> check for present==1
+            let path = device_path.join("present");
+            if path.exists() {
+                return read_file("battery", path).map_or(false, |x| x == "1");
+            }
         }
-        return false;
+        false
     }
 
-    fn refresh_device_info(&mut self) -> Result<()> {
-        if !self.is_available() {
-            // The user indicated that it's ok for this battery to be missing/go away
-            if self.allow_missing {
-                self.charge_full = None;
-                self.energy_full = None;
+    fn rescan_device(&mut self) -> Result<()> {
+        let sysfs_dir = "/sys/class/power_supply/";
+        let entries = read_dir(sysfs_dir).block_error(
+            "battery",
+            format!("failed to read {} directory", sysfs_dir).as_str(),
+        )?;
+        for dir in entries.flatten() {
+            if read_to_string(dir.path().join("type"))
+                .map(|t| t.trim() != "Battery")
+                .unwrap_or(false)
+            {
+                continue;
+            }
+            let dirname = dir.file_name();
+            let dirname = dirname.to_str().unwrap();
+            if self.device_regex.is_match(dirname) {
+                self.device_path = Some(PathBuf::from(sysfs_dir).join(dirname));
```

Comment:
Before assigning the path, it's availability should be checked. (I've implemented rescanning [like this](https://github.com/greshake/i3status-rust/blob/5f67435ef03167122385b39e6a822e096e214c4e/src/blocks/battery/sysfs.rs#L70-L96)).

Reply:
Does someone with `device = "BAT"` in his config want to match any of his batteries or just available ones?

For the default config I think it definitely makes sense to match the first _available_ device though.

With your implementation BATx / CMBx devices currently don't take precedence over e.g. hid devices anymore (as you mentioned above).

Should we change that back?
In the docs it states
```
sysfs: the first battery device found in /sys/class/power_supply, with "BATx" or "CMBx" entries taking precedence.
apcaccess: "localhost:3551"
upower: `DisplayDevice`
```

I'm not sure what makes more sense really, but not checking availability would be the faster implementation. 
(except for the default configuration, where this would be a breaking change)

Reply:
> Does someone with device = "BAT" in his config want to match any of his batteries or just available ones?

I think only available. What's the point of matching unavailable batteries? If there are two batteries: `BAT0` and `BAT1`, and only `BAT1` is available, and I have `device = "BAT."`, then I don't see why `BAT0` should be matched.

Furthermore, how is a battery with `present=0` different from a non-existing battery? I mean, some drivers just don't show removed batteries in the sysfs while some drivers indicate that it's not present.

> With your implementation BATx / CMBx devices currently don't take precedence over e.g. hid devices anymore

That's a just matter of default `device` value and is easy to change.